arcCommon/src/main/java/org/threadly/db/aurora/AuroraClusterMonitor.java

```diff
@@ -377,44 +445,128 @@ protected void checkAllServers() {
+        
+        if (monitor.isHealthy()) {
+          if (monitor.isMasterServer()) {
+            if (! as.equals(masterServer.get())) {
               // either already was the master, or now is
-              masterServer.set(p.getKey());
-              LOG.info("New master server: " + p.getKey());
+              masterServer.set(as);
+              LOG.info("New master server: " + as);
             }
           } else {
-            if (p.getKey().equals(masterServer.get())) {
+            if (as.equals(masterServer.get())) {
               // no longer a master, will need to find the new one
-              masterServer.compareAndSet(p.getKey(), null);
+              masterServer.compareAndSet(as, null);
               checkAllServers();  // make sure we find a new master ASAP
             }
-            if (! secondaryServers.contains(p.getKey())) {
-              secondaryServers.add(p.getKey());
+            
+            if (! replicaServers.contains(as)) {
+              replicaServers.add(as);
+
+              if (as.getWeight() > 0) {
+                if (updatedWeightedReplicaServers == null) {
+                  updatedWeightedReplicaServers = copyWeightedReplicaServers();
+                }
+                addWeightedServer(updatedWeightedReplicaServers, as);
+              }
             }
           }
         } else {
-          if (secondaryServers.remove(p.getKey())) {
-            // was secondary, so done
-            // TODO - removal of a secondary server can indicate it will become a new primary
+          if (replicaServers.remove(as)) {
+            if (as.getWeight() > 0) {
+              if (updatedWeightedReplicaServers == null) {
+                updatedWeightedReplicaServers = copyWeightedReplicaServers();
+              }
+              removeWeightedServer(updatedWeightedReplicaServers, as);
+            }
+            // TODO - removal of a replica server can indicate it will become a new primary
             //        How can we use this common behavior to recover quicker?
-          } else if (p.getKey().equals(masterServer.get())) {
+          } else if (as.equals(masterServer.get())) {
             // no master till we find a new one
-            masterServer.compareAndSet(p.getKey(), null);
+            masterServer.compareAndSet(as, null);
             checkAllServers();  // make sure we find a new master ASAP
           } // else, server is already known to be unhealthy, no change
         }
       }
+      
+      if (updatedWeightedReplicaServers != null) {
+        if (updatedWeightedReplicaServers.isEmpty()) {
+          weightedReplicaServers = Collections.emptyList();
+        } else if (updatedWeightedReplicaServers.size() == 1) {
+          weightedReplicaServers = Collections.singletonList(updatedWeightedReplicaServers.get(0));
+        } else {
+          updatedWeightedReplicaServers.trimToSize();
+          weightedReplicaServers = updatedWeightedReplicaServers;
```

Comment:
Given that this updates `weightedReplicaServers` all at once, wouldn't it be better to just do a `Collections.shuffle()` here and ditch the complex and inefficient random-insertion logic in `addWeightedServer()`?

Reply:
Benchmarking shows that the answer is "it depends".  If the weight is only one or two, this logic is ideal.  For higher weights you're right that a shuffle is more performant.  Part of what lead me to this is that we are guaranteed to have two well placed weights with this logic, where a shuffle is completely random.

I was thinking lower weights are probably more common, and that the overhead on this rare process is likely not a concern.  Thoughts?

Reply:
> two well placed weights

What does this mean?

Reply:
We choose a "random" server by going through the list of replica's linearly for efficiency.  I considered when writing the weight code that having a sever repeated means that this behavior will cause spikes as it sends repeated queries to one server before it finds the next one in the list.

So by well placed I mean "as far from each other as possible"